File: 14  Graphs/Learn/largest_island.java
class largest_island{
	public static void main (String args[]){
		int island[][] = {
                            {1, 0, 0, 1, 0},
                            {1, 0, 1, 0, 0},
                            {0, 0, 1, 0, 1},
                            {1, 0, 1, 1, 1},
                            {1, 0, 1, 1, 0}
                            };
        int res = 0;
        int count = 0;
        for(int i=0;i<island.length;i++){
        	for(int j=0;j<island[0].length;j++){
        		if(island[i][j]!=0){
        			count++;
        		res = Math.max(largest(island, i,j), res);
        		}
        		
        	}
        }
        System.out.println("largest island: "+res+" operations: "+count);
    }
    public static int largest(int graph[][], int i, int j){
    	if(i<0 || i>=graph.length || j<0 || j>=graph[0].length || graph[i][j]==0){
    		return 0;
    	}
    	int dx[] = {1,-1,0, 0};
    	int dy[] = {0, 0,1,-1};
    	int l = 0;
    	graph[i][j]=0;
    	for(int k =0;k<4;k++){

    		l += largest(graph,i+dx[k],j+dy[k]);
    		// max_l = Math.max(l, max_l);
    	}
    	return l+1;
    }
	
}

File: 14  Graphs/Revision/largest_island.cpp
#include<iostream>
#include<vector>

using namespace std;

int dfs(vector<vector<int>> &grid, int i, int j){
    if(grid[i][j]==0){
        return 0;
    }
    grid[i][j] = 0;

    int n = grid.size();
    int m = grid[0].size();
    int dx[] = {0,0,1,-1};
    int dy[] = {1,-1,0,0};

    int res = 0;

    for(int k =0; k<4;k++){
        int x = i + dx[k];
        int y = j + dy[k];
        if(x<0 || y<0 || x>=n || y>=m || grid[x][y]==0){
            continue;
        }
        res += dfs(grid, x,y);
    }
    return 1+res;
}

int largest_island(vector<vector<int>> grid){
    int n = grid.size();
    int m = grid[0].size();

    int largest = 0;

    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            largest = max(largest, dfs(grid, i, j));
        }
    }
    return largest;
}

int main(){
    vector<vector<int> > grid = {
                            {1, 0, 0, 1, 0},
                            {1, 0, 1, 0, 0},
                            {0, 0, 1, 0, 1},
                            {1, 0, 1, 1, 1},
                            {1, 0, 1, 1, 0}
                            };
    
    cout<<"Largest Island: "<< largest_island(grid) <<endl;


    return 0;
}